File: go-arena/internal/game/spatial.go
package game

import "math"

// SpatialGrid provides O(1) proximity queries using a fixed-size grid.
type SpatialGrid struct {
	cellSize  float64
	cells     map[[2]int]map[string]bool // grid cell -> set of entity IDs
	positions map[string]Vec2            // entity_id -> position
}

// NewSpatialGrid creates a new spatial grid with the given cell size.
func NewSpatialGrid(cellSize float64) *SpatialGrid {
	return &SpatialGrid{
		cellSize:  cellSize,
		cells:     make(map[[2]int]map[string]bool),
		positions: make(map[string]Vec2),
	}
}

// cellKey returns the grid cell coordinates for a world position.
func (g *SpatialGrid) cellKey(x, y float64) [2]int {
	return [2]int{
		int(math.Floor(x / g.cellSize)),
		int(math.Floor(y / g.cellSize)),
	}
}

// Insert adds an entity to the grid at the given position.
func (g *SpatialGrid) Insert(id string, pos Vec2) {
	key := g.cellKey(pos.X(), pos.Y())
	if g.cells[key] == nil {
		g.cells[key] = make(map[string]bool)
	}
	g.cells[key][id] = true
	g.positions[id] = pos
}

// Remove removes an entity from the grid.
func (g *SpatialGrid) Remove(id string) {
	pos, ok := g.positions[id]
	if !ok {
		return
	}
	key := g.cellKey(pos.X(), pos.Y())
	if cell := g.cells[key]; cell != nil {
		delete(cell, id)
		if len(cell) == 0 {
			delete(g.cells, key)
		}
	}
	delete(g.positions, id)
}

// Update moves an entity to a new position in the grid. Only reassigns cells
// if the entity has moved to a different cell.
func (g *SpatialGrid) Update(id string, pos Vec2) {
	oldPos, ok := g.positions[id]
	if !ok {
		g.Insert(id, pos)
		return
	}

	oldKey := g.cellKey(oldPos.X(), oldPos.Y())
	newKey := g.cellKey(pos.X(), pos.Y())

	if oldKey != newKey {
		// Remove from old cell
		if cell := g.cells[oldKey]; cell != nil {
			delete(cell, id)
			if len(cell) == 0 {
				delete(g.cells, oldKey)
			}
		}
		// Insert into new cell
		if g.cells[newKey] == nil {
			g.cells[newKey] = make(map[string]bool)
		}
		g.cells[newKey][id] = true
	}

	g.positions[id] = pos
}

// QueryRadius returns all entity IDs within the given Euclidean distance of pos.
func (g *SpatialGrid) QueryRadius(pos Vec2, radius float64) []string {
	// Determine the range of cells that could overlap the radius bounding box.
	minCX := int(math.Floor((pos.X() - radius) / g.cellSize))
	maxCX := int(math.Floor((pos.X() + radius) / g.cellSize))
	minCY := int(math.Floor((pos.Y() - radius) / g.cellSize))
	maxCY := int(math.Floor((pos.Y() + radius) / g.cellSize))

	r2 := radius * radius
	var result []string

	for cx := minCX; cx <= maxCX; cx++ {
		for cy := minCY; cy <= maxCY; cy++ {
			cell := g.cells[[2]int{cx, cy}]
			if cell == nil {
				continue
			}
			for id := range cell {
				ePos := g.positions[id]
				dx := ePos.X() - pos.X()
				dy := ePos.Y() - pos.Y()
				if dx*dx+dy*dy <= r2 {
					result = append(result, id)
				}
			}
		}
	}
	return result
}

// GetPosition returns the stored position of an entity and whether it exists.
func (g *SpatialGrid) GetPosition(id string) (Vec2, bool) {
	pos, ok := g.positions[id]
	return pos, ok
}

// Clear removes all entities from the grid.
func (g *SpatialGrid) Clear() {
	g.cells = make(map[[2]int]map[string]bool)
	g.positions = make(map[string]Vec2)
}

File: server/game/spatial.py
"""Spatial grid for efficient proximity queries in the arena."""

import math
from collections import defaultdict

from server.config import settings


class SpatialGrid:
    """Grid-based spatial index for O(1) cell lookups and fast radius queries.

    Divides the arena into cells. Each entity maps to one cell based on position.
    Radius queries only check neighboring cells that could contain matches.
    """

    def __init__(
        self,
        width: int | None = None,
        height: int | None = None,
        cell_size: int | None = None,
    ) -> None:
        """Initialize spatial grid.

        Args:
            width: Arena width in units. Defaults to config value.
            height: Arena height in units. Defaults to config value.
            cell_size: Grid cell size in units. Defaults to config value.
        """
        self.width = width or settings.game.arena_width
        self.height = height or settings.game.arena_height
        self.cell_size = cell_size or settings.game.spatial_cell_size
        self.cols = math.ceil(self.width / self.cell_size)
        self.rows = math.ceil(self.height / self.cell_size)

        # cell (cx, cy) -> set of entity_ids
        self._cells: dict[tuple[int, int], set[str]] = defaultdict(set)
        # entity_id -> (cx, cy) current cell
        self._entity_cells: dict[str, tuple[int, int]] = {}
        # entity_id -> (x, y) exact position
        self._positions: dict[str, tuple[float, float]] = {}

    def _get_cell(self, x: float, y: float) -> tuple[int, int]:
        """Get grid cell coordinates for a world position."""
        cx = min(int(x / self.cell_size), self.cols - 1)
        cy = min(int(y / self.cell_size), self.rows - 1)
        return (max(0, cx), max(0, cy))

    def insert(self, entity_id: str, x: float, y: float) -> None:
        """Add an entity to the grid at the given position.

        If the entity already exists it is removed first to prevent ghost entries.
        """
        if entity_id in self._entity_cells:
            self.remove(entity_id)
        cell = self._get_cell(x, y)
        self._cells[cell].add(entity_id)
        self._entity_cells[entity_id] = cell
        self._positions[entity_id] = (x, y)

    def remove(self, entity_id: str) -> None:
        """Remove an entity from the grid."""
        cell = self._entity_cells.pop(entity_id, None)
        if cell is not None:
            self._cells[cell].discard(entity_id)
        self._positions.pop(entity_id, None)

    def update(self, entity_id: str, x: float, y: float) -> None:
        """Move an entity to a new position, updating its cell if needed."""
        new_cell = self._get_cell(x, y)
        old_cell = self._entity_cells.get(entity_id)

        if old_cell != new_cell:
            if old_cell is not None:
                self._cells[old_cell].discard(entity_id)
            self._cells[new_cell].add(entity_id)
            self._entity_cells[entity_id] = new_cell

        self._positions[entity_id] = (x, y)

    def query_radius(self, x: float, y: float, radius: float) -> list[str]:
        """Find all entity IDs within radius of (x, y).

        Only checks cells that could contain entities within the radius.
        """
        radius_sq = radius * radius
        # Determine cell range to check
        min_cx = max(0, int((x - radius) / self.cell_size))
        max_cx = min(self.cols - 1, int((x + radius) / self.cell_size))
        min_cy = max(0, int((y - radius) / self.cell_size))
        max_cy = min(self.rows - 1, int((y + radius) / self.cell_size))

        result: list[str] = []
        for cx in range(min_cx, max_cx + 1):
            for cy in range(min_cy, max_cy + 1):
                for eid in self._cells.get((cx, cy), set()):
                    pos = self._positions[eid]
                    dx = pos[0] - x
                    dy = pos[1] - y
                    if dx * dx + dy * dy <= radius_sq:
                        result.append(eid)
        return result

    def query_cell(self, cx: int, cy: int) -> list[str]:
        """Get all entity IDs in a specific grid cell."""
        return list(self._cells.get((cx, cy), set()))

    def get_position(self, entity_id: str) -> tuple[float, float] | None:
        """Get the stored position of an entity."""
        return self._positions.get(entity_id)

    def clear(self) -> None:
        """Remove all entities from the grid."""
        self._cells.clear()
        self._entity_cells.clear()
        self._positions.clear()
